File: 0-frameworks-in-backend/1-java/src/main/java/com/example/dog/DogController.java
package com.example.dog;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class DogController {
    private final DogService dogService;

    public DogController(DogService dogService) {
        this.dogService = dogService;
    }

    @GetMapping("/dogs/spy")
    public DogService.SpyReport spy() {
        return dogService.spyReport();
    }

    @GetMapping("/dogs/cats-via-di")
    public DogService.CatsView catsViaDI() {
        return dogService.catsViaDI();
    }
}

File: 0-frameworks-in-backend/0-typescript/src/dog/dog.controller.ts
/**
 * Controller handling `/dogs/*` endpoint — cross-module dependency demo.
 */
import {
    Controller,
    Get,
} from "@nestjs/common"
import {
    DogService,
} from "./dog.service"

@Controller("dogs")
export class DogController {
    /**
     * Inject service so controller stays as an orchestration layer.
     */
    constructor(private readonly dogService: DogService) {}

    /**
     * Handle `GET /dogs/spy` for cross-module dependency test.
     */
    @Get("spy")
    spy(): { mission: string; dependency: string; status: string } {
        // Keep business logic out of controller for easier testing and maintenance.
        return this.dogService.getSpyReport()
    }

    /**
     * Handle `GET /dogs/cats-via-di` — demonstrates cross-module DI exposing
     * the full `CatService` API surface, not just a single hint method.
     */
    @Get("cats-via-di")
    borrowAllCats(): { dog: string; borrowedCats: Array<{ id: number; name: string }> } {
        return this.dogService.borrowAllCats()
    }
}
